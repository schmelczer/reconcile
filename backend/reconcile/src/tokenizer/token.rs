#[derive(Debug, Clone)]
pub struct Token<T>
where
    T: PartialEq + Clone,
{
    normalised: T,
    original: String,
}

impl From<&str> for Token<String> {
    fn from(s: &str) -> Self {
        Token {
            normalised: s.to_string(),
            original: s.to_string(),
        }
    }
}

impl<T> Token<T>
where
    T: PartialEq + Clone,
{
    pub fn new(normalised: T, original: String) -> Self {
        Token {
            normalised,
            original,
        }
    }

    pub fn original(&self) -> &str {
        &self.original
    }

    pub fn normalised(&self) -> &T {
        &self.normalised
    }

    pub fn get_original_length(&self) -> usize {
        self.original.chars().count()
    }
}

impl<T> PartialEq for Token<T>
where
    T: PartialEq + Clone,
{
    fn eq(&self, other: &Self) -> bool {
        self.normalised == other.normalised
    }
}
