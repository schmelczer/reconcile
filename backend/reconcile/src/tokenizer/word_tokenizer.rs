use super::token::Token;

/// Splits on word boundaries creating alternating words and whitespaces with
/// the whitesspaces getting unique IDs.
///
/// ## Example
///
/// "Hi there!" -> ["Hi", " " ", "there!"]
pub fn word_tokenizer(text: &str) -> Vec<Token<String>> {
    let mut result: Vec<Token<String>> = Vec::new();

    let mut previous_boundary_index = 0;
    let mut previous_char_is_whitespace = text.chars().next().map_or(true, |c| c.is_whitespace());

    for (i, c) in text.char_indices() {
        let is_current_char_whitespace = c.is_whitespace();
        if previous_char_is_whitespace != is_current_char_whitespace {
            result.push(text[previous_boundary_index..i].into());
            previous_boundary_index = i;
        }

        previous_char_is_whitespace = is_current_char_whitespace;
    }

    if previous_boundary_index < text.len() {
        result.push(text[previous_boundary_index..].into());
    }

    if result.is_empty() {
        return result;
    }

    for i in 0..result.len() - 1 {
        if result[i].original().chars().all(|c| c.is_whitespace()) {
            result[i].normalised = result[i].normalised().to_owned() + result[i + 1].original()
        }
    }

    result
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::*;

    #[test]
    fn test_with_snapshots() {
        assert_debug_snapshot!(word_tokenizer("Hi there!"));

        assert_debug_snapshot!(word_tokenizer(""));

        assert_debug_snapshot!(word_tokenizer(" what? "));

        assert_debug_snapshot!(word_tokenizer(" hello, \nwhere are you?"));
    }
}
