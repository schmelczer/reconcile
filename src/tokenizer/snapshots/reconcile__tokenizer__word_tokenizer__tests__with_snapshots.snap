---
source: reconcile/src/tokenizer/word_tokenizer.rs
expression: "word_tokenizer(\"Hi there!\")"
snapshot_kind: text
---
[
    Token {
        normalised: "Hi",
        original: "Hi",
        is_left_joinable: true,
        is_right_joinable: true,
    },
    Token {
        normalised: " there!",
        original: " ",
        is_left_joinable: true,
        is_right_joinable: true,
    },
    Token {
        normalised: "there!",
        original: "there!",
        is_left_joinable: true,
        is_right_joinable: true,
    },
]
